// igdb-api-node/index.js
// Generated by CoffeeScript 1.10.0
(function() {
  var config, get, request;

  request = require('request');

  config = require('./config');

  get = function(url, callback, opts) {
    var filter, filterValue, optUrl, options, param, paramValue, errCounter = 0;
    if (opts == null) {
      opts = false;
    }
    url = config.endpoint + url;
    if (opts) {
      optUrl = [];
      for (param in opts) {
        paramValue = opts[param];
        if (param === "filters") {
          for (filter in paramValue) {
            filterValue = paramValue[filter];
            optUrl.push("filters[" + filter + "]=" + filterValue);
          }
        } else {
          optUrl.push(param + "=" + paramValue);
        }
      }
      url += "?" + optUrl.join('&');
    }
    options = {
      url: url,
      headers: {
        'Accept': 'application/json',
        'Authorization': 'Token token="' + config.apikey + '"'
      }
    };
    return request(options, function(error, response, body) {
      if (error) {
        throw error;
      }
      if (response.statusCode === 200) {
        if (callback) {
          errCounter = 0
          return callback(JSON.parse(body));
        }
      } else {
        errCounter++
        console.log(64, errCounter)
        if (errCounter > 10) {
          throw response.statusCode;
        } else {
          get(url, callback, opts)
        }
      }
    });
  };

  module.exports = {
    games: {
      index: function(opts, callback) {
        return get('games', callback, opts);
      },
      get: function(id, callback) {
        return get('games/' + id, callback);
      },
      meta: function(callback) {
        return get('games/meta', callback);
      },
      search: function(opts, callback) {
        return get('games/search', callback, opts);
      }
    },
    companies: {
      index: function(opts, callback) {
        return get('companies', callback, opts);
      },
      get: function(id, callback) {
        return get('companies/' + id, callback);
      },
      meta: function(callback) {
        return get('companies/meta', callback);
      },
      games: function(opts, id, callback) {
        return get('companies/' + id + '/games', callback, opts);
      }
    },
    people: {
      index: function(opts, callback) {
        return get('people', callback, opts);
      },
      get: function(id, callback) {
        return get('people/' + id, callback);
      },
      meta: function(callback) {
        return get('people/meta', callback);
      },
      games: function(opts, id, callback) {
        return get('people/' + id + '/games', callback, opts);
      },
      titles: function(opts, id, callback) {
        return get('people/' + id + '/titles', callback, opts);
      }
    }
  };

}).call(this);
